Components/ hooks contributing guide (#2820)
clerkinc/clerk-docs: contributing/CONTRIBUTING-COMPONENTS-HOOKS.md

## Hooks

All existing hooks are listed in the [Hooks Reference overview](https://clerk.com/docs/nextjs/reference/hooks/overview), and live under `docs/reference/hooks` within our docs.

### How hook documentation works

A hook page _usually_ includes the following:

- A description of the hook.
- A `Parameters` section, listing the parameters the hook accepts.
- A `Returns` section, listing the properties the hook returns.
- SDK-specific code examples showing how to use the hook, with explanations.
- Any additional information.

Unlike component documentation, hook documentation is split between two sources. This is intentional, and reflects how we handle type information versus written content. Historically, many hook pages were fully generated from JSDoc + Typedoc, including code examples. However, adding examples for multiple SDKs through Typedoc required significant restructuring of the `clerk/javascript` repo, so we shifted to a hybrid model.

Each hook page contains two distinct layers:

1. Typedoc-generated content

> The parameters, return types, and type definitions for each hook are autogenerated from [JSDoc comments](https://jsdoc.app/about-getting-started) in the `clerk/javascript` repo. Typedoc converts these comments into MDX output, which lives under `/clerk-typedoc` in `clerk-docs`. These MDX files are embedded via the [<Typedoc />](https://github.com/clerk/clerk-docs/blob/main/CONTRIBUTING.md#typedoc-) component. You can learn more about this process in [this section](https://github.com/clerk/javascript/blob/main/docs/CONTRIBUTING.md#authoring-typedoc-information).

Reviewer: ```suggestion
The parameters, return types, and type definitions for each hook are autogenerated from [JSDoc comments](https://jsdoc.app/about-getting-started) in the `clerk/javascript` repo. Typedoc converts these comments into MDX output, which lives in `clerk-docs` in the `clerk-typedoc/` folder. These MDX files are embedded via the [<Typedoc />](https://github.com/clerk/clerk-docs/blob/main/CONTRIBUTING.md#typedoc-) component. You can learn more about this process in [this section](https://github.com/clerk/javascript/blob/main/docs/CONTRIBUTING.md#authoring-typedoc-information).
```

- The [clerk-docs](https://github.com/clerk/clerk-docs) repository, when adding or updating explanations, SDK-specific code examples, or any other written content.

If you need to update **parameters or return types** for a hook, the changes must be made in the `clerk/javascript` repo:

- Most hooks live under the `packages/shared` folder, except for three hooks - `useAuth`, `useSignIn`, and `useSignup` - who live under the `packages/react` folder.
- Each hook's source file includes interfaces or types made visible via JSDoc comments.
- You can add, remove, or update properties and descriptions directly in these JSDoc comments.

When embedding Typedoc output for a hook in the docs, the package you import from inside the `clerk-typedoc` folder depends on where that hook is exported from in the `clerk/javascript` repo. Use the following mapping:

- If the hook is exported from `packages/react/src/hooks/index.ts`, **import its Typedoc output from the `clerk-react` package**, with the exception of `useOrganization`, `useOrganizationList`, and `useReverification`.

> - If the hook is exported from `packages/shared/src/react/hooks/index.ts`, **import its Typedoc output from the `shared` package**.

Reviewer: maybe throw examples for each of these? to show what you mean by "import its typedoc output"

like
```
<Include src="clerk-react/<FILENAME>.mdx />
```

# Contributing to Clerk's hooks and components documentation

> Component and hook documentation includes SDK-specific code examples and supporting content across all supported SDKs to ensure the best user experience. This means every component and hook must have code examples for each SDK it supports. If you're adding a new component or hook, or updating an existing one, the sections below outline the process for each.

Reviewer 2: A slightly more concise intro. You can take it or leave it.

```suggestion
Component and hook documentation includes SDK-specific code examples and supporting content for all supported SDKs to ensure the best user experience. Each component and hook must have code examples for each supported SDK. If you’re adding a new component or updating an existing one, the sections below outline the process.
```

<details open="open">
<summary><strong>Table of contents</strong></summary>

- [Contributing to Clerk's hooks and components documentation](#contributing-to-clerks-hooks-and-components-documentation)
  - [Components](#components)
    - [How component documentation works](#how-component-documentation-works)
    - [How to update component documentation](#how-to-update-component-documentation)
  - [Hooks](#hooks)
    - [How hook documentation works](#how-hook-documentation-works)
    - [How to update hook documentation](#how-to-update-hook-documentation)

</details>

## Components

> All existing components are listed in the [Component Reference overview](https://clerk.com/docs/nextjs/reference/components/overview), and live under `docs/reference/components` within our docs. Each component belongs to a specific category apart from `<ClerkProvider>`, and these categories determine how components are organized in both the sidebar and the reference overview. **It's therefore important to ensure that the ordering in the sidebar matches the ordering in the reference overview**.

Reviewer 2: ```suggestion
All existing components are listed in the [Component Reference overview](https://clerk.com/docs/nextjs/reference/components/overview), and live under [`docs/reference/components`](/docs/reference/components) within our docs. Each component belongs to a specific category (except for `<ClerkProvider>`), and these categories determine how components are organized in both the sidebar and the reference overview. **Therefore, it's important to ensure that the ordering in the sidebar matches the ordering in the reference overview**.
```

Reviewer: oh to add to michael's suggestion, lets remove the sdk-scoping from the link:
```suggestion
All existing components are listed in the [Component Reference overview](https://clerk.com/docs/reference/components/overview), and live under [`docs/reference/components`](/docs/reference/components) within our docs. Each component belongs to a specific category (except for `<ClerkProvider>`), and these categories determine how components are organized in both the sidebar and the reference overview. **Therefore, it's important to ensure that the ordering in the sidebar matches the ordering in the reference overview**.
```

### How component documentation works

A component page _usually_ includes the following:

- An image of the component at the top.
- A description of the component.
- An SDK-specific code example showing how to use the component, with an explanation.
- A list of any optional properties the component accepts, rendered via the [`<Properties>`](https://github.com/clerk/clerk-docs/blob/main/CONTRIBUTING.md#properties-1) component.
- Any additional information.

### How to update component documentation

**All component documentation is directly maintained in this repository.**

If you have to make any changes to the components documentation, here are two common scenarios you may encounter:

- If **you need to add a new component** to the docs:

  - Add the component to the [Component Reference overview](https://clerk.com/docs/nextjs/reference/components/overview) under its corresponding category.
  - Add the component to the sidebar under the same category.

>   - Include an image of the component at the top of the page, in both SVG and PNG formats. You can request image assets from the Design team by contacting them directly or by creating a ticket in their [Linear board](https://linear.app/clerk/team/DSN/all). To learn how to add images to the docs, see this [section](./CONTRIBUTING.md#images-and-static-assets).

Reviewer 2: ```suggestion
  - Include an image of the component at the top of the page, in both SVG and PNG formats. You can request image assets from the Design team by contacting them directly or by creating a ticket in their [Linear board](https://linear.app/clerk/team/DSN/all). Learn [how to add images to the docs](./CONTRIBUTING.md#images-and-static-assets).
```
